From 4312fbcb1d8d58f61d127617f5d5445a4aa0e54f Mon Sep 17 00:00:00 2001
From: Jens Nyberg I <jens.i.nyberg@ericsson.com>
Date: Wed, 30 Nov 2011 12:01:09 -0500
Subject: [PATCH 1/3] Fix ipcalc_netmask

This resolves: 

  https://bugzilla.redhat.com/show_bug.cgi?id=848722

1) make the bitmask creation simpler to understand.

2) effectively increase string length for inet_ntop result by 1 (it
   already takes the terminating NULL into account).
(cherry picked from commit d340f2dfcd6461c9743dccdabe3b610f5fbc8fe8)
---
 src/xslt_ext.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/xslt_ext.c b/src/xslt_ext.c
index 2bdc771..596ef77 100644
--- a/src/xslt_ext.c
+++ b/src/xslt_ext.c
@@ -72,10 +72,10 @@ static void ipcalc_netmask(xmlXPathParserContextPtr ctxt, int nargs) {
     struct in_addr netmask;
     xmlChar netmask_str[16];
 
-    netmask.s_addr = htonl(~((1 << (32 - prefix)) - 1));
+    netmask.s_addr = htonl(~(0xffffffffu >> prefix));
 
     if (! inet_ntop(AF_INET, &netmask,
-                    (char *) netmask_str, sizeof(netmask_str) - 1)) {
+                    (char *) netmask_str, sizeof(netmask_str))) {
         xsltTransformError(xsltXPathGetTransformContext(ctxt), NULL, NULL,
                     "ipcalc:netmask: internal error: inet_ntop failed");
         goto error;
-- 
1.7.11.7

